osquery/events/linux/syslog.cpp

```diff
@@ -0,0 +1,209 @@
+
+  return Status(0, "OK");
+}
+
+Status SyslogEventPublisher::createPipe(const std::string& path) {
+  if (mkfifo(FLAGS_syslog_pipe_path.c_str(), kPipeMode) != 0) {
+    return Status(1, "Error in mkfifo: " + std::string(strerror(errno)));
+  }
+
+  // Explicitly set the permissions since the umask will effect the
+  // permissions created by mkfifo
+  if (chmod(FLAGS_syslog_pipe_path.c_str(), kPipeMode) != 0) {
+    return Status(1, "Error in chmod: " + std::string(strerror(errno)));
+  }
+
+  // Try to set the group so that rsyslog will be able to write to the pipe
+  struct group* group = getgrnam(kPipeGroupName.c_str());
+  if (group == nullptr) {
+    VLOG(1) << "No group " << kPipeGroupName
+            << " found. Not changing group for the pipe.";
+    return Status(0, "OK");
+  }
+  if (chown(FLAGS_syslog_pipe_path.c_str(), -1, group->gr_gid) == -1) {
+    return Status(1, "Error in chown to group " + kPipeGroupName + ": " +
+                         std::string(strerror(errno)));
+  }
+  return Status(0, "OK");
+}
+
+Status SyslogEventPublisher::lockPipe(const std::string& path) {
+  lockFd_ = open(path.c_str(), O_NONBLOCK);
+  if (lockFd_ == -1) {
+    return Status(1, "Error in open for locking pipe: " +
+                         std::string(strerror(errno)));
+  }
+  if (flock(lockFd_, LOCK_EX | LOCK_NB) != 0) {
+    lockFd_ = -1;
+    return Status(1, "Unable to acquire pipe lock: " +
+                         std::string(strerror(errno)));
+  }
+  return Status(0, "OK");
+}
+
+void SyslogEventPublisher::unlockPipe() {
+  if (lockFd_ != -1) {
+    if (flock(lockFd_, LOCK_UN) != 0) {
+      LOG(WARNING) << "Error unlocking pipe: " << std::string(strerror(errno));
+    }
+  }
+}
+
+Status SyslogEventPublisher::run() {
+  // This run function will be called by the event factory with ~100ms pause
+  // (see InterruptableRunnable::pause()) between runs. In case something goes
+  // weird and there is a huge amount of input, we limit how many logs we
+  // take in per run to avoid pegging the CPU.
+  for (size_t i = 0; i < kMaxLogsPerRun; ++i) {
+    if (readStream_.rdbuf()->in_avail() == 0) {
+      // If there is no pending data, we have flushed everything and can wait
+      // until the next time EventFactory calls run(). This also allows the
+      // thread to join when it is stopped by EventFactory.
+      return Status(0, "OK");
+    }
+    std::string line;
+    std::getline(readStream_, line);
+    auto ec = createEventContext();
+    Status status = populateEventContext(line, ec);
+    if (status.ok()) {
+      fire(ec);
+      if (errorCount_ > 0) {
+        --errorCount_;
+      }
+    } else {
+      LOG(ERROR) << status.getMessage() << " in line: " << line;
+      ++errorCount_;
+      if (errorCount_ >= kErrorThreshold) {
+        return Status(1, "Too many errors in syslog parsing.");
+      }
+    }
+  }
+  return Status(0, "OK");
+}
+
+void SyslogEventPublisher::tearDown() { unlockPipe(); }
```

Comment:
Do you only want to unlock access to the FIFO?

Is the intention to leave the FIFO open, expecting rsyslog to continue to write log lines, and for osquery to come back online 'soon'?